distro/src/bin/hz

```diff
@@ -20,19 +20,75 @@ PRGDIR=`cd "$PRGDIR"; pwd`
 # set the parent directory as HAZELCAST_HOME.
 export HAZELCAST_HOME=${HAZELCAST_HOME:-$(cd "$PRGDIR/.." >/dev/null; pwd -P)}
 
-if [ $JAVA_HOME ] ; then
-    RUN_JAVA=$JAVA_HOME/bin/java
-else
-    RUN_JAVA=`command -v java 2>/dev/null`
+# common.sh needs it
+SCRIPT_DIR="$PRGDIR"
+
+. "$PRGDIR"/common.sh
+
+
+if [ "${PROMETHEUS_PORT}" ]; then
+    echo "Prometheus enabled on port ${PROMETHEUS_PORT}"
+    PROMETHEUS="-javaagent:${HAZELCAST_HOME}/lib/jmx_prometheus_javaagent-0.14.0.jar=${PROMETHEUS_PORT}:${HAZELCAST_HOME}/config/jmx_agent_config.yaml"
 fi
 
-if [ -z $RUN_JAVA ] ; then
-    echo "Error: JRE not found. Please install JRE 8 or higher in your PATH or set JAVA_HOME appropriately"
-    exit 1
+readJvmOptionsFile "jvm.options"
+
+if [ "$HAZELCAST_LICENSE_KEY" ]; then
+    LICENSING_OPT="-Dhazelcast.enterprise.license.key=${HAZELCAST_LICENSE_KEY}"
 fi
 
-export CLASSPATH="$HAZELCAST_HOME/lib/hazelcast-command-line-CLI_VERSION.jar:$HAZELCAST_HOME/lib/hazelcast-HZ_VERSION.jar:$HAZELCAST_HOME/bin/download:$HAZELCAST_HOME/bin/download/hazelcast-download.properties"
+JAVA_OPTS_ARRAY=(\
+$JDK_OPTS \
+"-Dhazelcast.logging.type=log4j2" \
+"-Dlog4j.configurationFile=file:$HAZELCAST_HOME/config/log4j2.properties" \
+"-Dhz.home=$HAZELCAST_HOME" \
+"-Dhazelcast.config=$HAZELCAST_HOME/config/hazelcast.xml" \
+"-Djet.custom.lib.dir=$HAZELCAST_HOME/custom-lib" \
+$PROMETHEUS \
+$LICENSING_OPT\
+$JVM_OPTIONS \
+$JAVA_OPTS \
+)
 
-HZ_CMD="${RUN_JAVA} -cp ${CLASSPATH} ${JAVA_OPTS} -Dhazelcast.commandline.workingdirectory=${HAZELCAST_HOME} com.hazelcast.commandline.HazelcastCommandLine"
+if [ -z "$LOGGING_LEVEL" ]; then
+    export LOGGING_LEVEL="INFO"
+fi
 
-${HZ_CMD} "$@"
+if [ "$(arch)" == "s390x" ]; then
+  export LOGGING_PATTERN="%d [%highlight{\${LOG_LEVEL_PATTERN:-%5p}}{FATAL=red, ERROR=red, WARN=yellow, INFO=green, DEBUG=magenta}][%style{%t{1.}}{cyan}] [%style{%-10c}{blue}]: %m%n"
+else
+  export LOGGING_PATTERN="%d [%highlight{\${LOG_LEVEL_PATTERN:-%5p}}{FATAL=red, ERROR=red, WARN=yellow, INFO=green, DEBUG=magenta}] [%style{%t{1.}}{cyan}] [%style{%c{1.}}{blue}]: %m%n"
+fi
+
+
+CLASSPATH="$HAZELCAST_HOME/lib:$HAZELCAST_HOME/bin/user-lib:$HAZELCAST_HOME/bin/user-lib/*"
+for filename in "$HAZELCAST_HOME"/lib/*; do
+  case "$filename" in
+    picocli*)
+      ;;
+    *)
+      CLASSPATH="$CLASSPATH:$filename"
+      ;;
+  esac
+done
+
+J_OPTS=""
+for arg in "${JAVA_OPTS_ARRAY[@]}"
+do
+  if [ "$J_OPTS" = "" ]; then
```

Comment:
I see in `hz-start` we also added `:$HAZELCAST_HOME/lib/*:` to the `CLASSPATH` but I guess here we just want to include picocli. Is that right, and how come?

Also, minor alignment - how about printing the options on startup like we do with `hz-start`?
```shell
########################################
# JAVA=/usr/bin/java
# JAVA_OPTS=--add-modules java.se --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED -Dhazelcast.logging.type=log4j2 -Dlog4j.configurationFile=file:/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/config/log4j2.properties -Dhazelcast.config=/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/config/hazelcast.xml -Djet.custom.lib.dir=/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/custom-lib
# CLASSPATH=:/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/lib:/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/lib/*:/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/bin/user-lib:/Users/matko/workspace/checkouts/hazelcast/distribution/target/hazelcast-5.0-SNAPSHOT/bin/user-lib/*
########################################
```

Reply:
Actually everything from `$HAZELCAST_HOME/lib/*` except picocli is added to the `CLASSPATH`, and the reason is CLI uses different version of picocli and it creates a conflict.
```shell
CLASSPATH="$HAZELCAST_HOME/lib:$HAZELCAST_HOME/bin/user-lib:$HAZELCAST_HOME/bin/user-lib/*"
for filename in "$HAZELCAST_HOME"/lib/*; do
  case "$filename" in
    picocli*)
      ;;
    *)
      CLASSPATH="$CLASSPATH:$filename"
      ;;
  esac
done
```
Thank you for the suggestion, will add printing options to the console on startup

Reply:
Regarding printing options/classpath to the console I tried it locally, but it is visually unpleasant. For example, when user runs `hz` it prints multiline class path/options to the console and then shows `--help`. So one option would be to do the printing only when `hz start` is invoked

Reply:
> Regarding printing options/classpath to the console I tried it locally, but it is visually unpleasant. For example, when user runs hz it prints multiline class path/options to the console and then shows --help. So one option would be to do the printing only when hz start is invoked

Yes, sure. It was just a suggestion to align the behaviour with `hz-start`. I see you added it so let me try and see how it looks.

Reply:
Ok, now I see :D
```
########################################
# JAVA=/usr/bin/java
# JAVA_OPTS=--add-modules java.se --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED -Dhazelcast.logging.type=log4j2 -Dlog4j.configurationFile=file:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/config/log4j2.properties -Dhazelcast.config=/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/config/hazelcast.xml -Djet.custom.lib.dir=/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/custom-lib
# CLASSPATH=/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/bin/user-lib:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/bin/user-lib/*:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/cache-api-1.1.1.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-3-connector-common-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-3-connector-interface-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-command-line-5.2021.08-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-download.properties:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-hibernate53-2.1.1.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-avro-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-cdc-debezium-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-cdc-mysql-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-cdc-postgres-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-csv-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-elasticsearch-7-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-files-azure-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-files-gcs-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-files-s3-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-grpc-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-hadoop-all-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-kafka-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-kinesis-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-protobuf-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-python-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-jet-s3-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-sql-5.0-SNAPSHOT.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/hazelcast-wm-4.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/jansi-2.1.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/jline-reader-3.19.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/jline-terminal-3.19.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/jline-terminal-jansi-3.19.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/jmx_prometheus_javaagent-0.14.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/log4j-api-2.14.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/log4j-core-2.14.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/log4j-slf4j-impl-2.14.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/picocli-3.9.0.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/lib/slf4j-api-1.7.30.jar:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/bin/download:/Users/matko/workspace/checkouts/hazelcast-command-line/distro/build/dist/hazelcast-5.0-SNAPSHOT/bin/download/hazelcast-download.properties
########################################
```

Ok, let's add it. I guess it might be useful when starting, to see what the instance is started with and if people complain we'll remove it.

Reply:
BTW thanks for the explanation around the classpaths, I now see my linux foo was too weak or I was too blind :)